### ISSUE TITLE
Decklink audio output is broken with NTSC or 59.94 setup

### ISSUE THREAD

**Issue by sendust** [CasparCG/server#1319]
### Expected behaviour
Test Tone singal playback should be clean

### Current behaviour
Test Tone singal playback is not clean when casparCG is setup with Decklink NTSC or 1080i5994
Playback with PAL or 50, 60 configuration is fine.

### Steps to reproduce
1. Create test tone signal with ffmpeg
  ffmpeg -f lavfi -i testsrc2=rate=29.97:duration=60:size=1920x1080 -f lavfi -i sine=frequency=1000:sample_rate=48000:duration=60 -af volume=-30dB -c:v mpeg2video -b:v 20000k -c:a pcm_s16le tone.mov
2. setup CasparCG with Decklink, NTSC
3. playback generated test signal
  play 1-1 tone
4. Listen HDSDI or AES audio from Decklink (Volume up)
5. You can hear abnormal tone sound with NTSC or 5994 setup

### Environment
* Commit: d442a8f84f68838f91c82f5f57c9634e84f7a23a and later
* Server version: [v 2.3.0]
* Operating system: [Windows 10]
 Decklink Driver [10.11.4 ~ 11.5.1]

with PAL or 50, 60 configuration, tone signal playback is perfect

You can check recorded decklink output here

Normal [PAL, 6000]
https://youtu.be/77crz2iHc5g
https://youtu.be/yaKXQ8g2YE0

Abnormal [NTSC, 5994]
https://youtu.be/ftGc4xQ8OgE
https://youtu.be/Xvi4Weer888

**Comment by hamuchen** [comment]
I can confirm this issue. Any updates, because the linked Issue #1326 seems fixed?

We are stuck on 2.0.7 because of the audio issues with every version after that with 5994.

Cheers,
hamu

**Comment by dedicatedbroadcastsolutions** [comment]
I am having the audio crackle issue with 2.4.2 stable at i5994 and p2997 rates.  We need some fix to this or this program becomes unusable for broadcast without frame rate converters in hardware.  This is effecting both screen and decklink outputs.

**Comment by TomKaltz** [comment]
@Julusian @ronag I'm wondering if the issues people are having with 1080i5994 are due to the strange cadence, currently: 801, 801, 801, 800, 801, 801, 801, 800, 801, 801

Is there any reason we aren't using a 5 frame cadence run like SMPTE recommends?  They mention a cadence run of 800, 802, 800, 802, 800.

Could we also do something like 801, 801, 801, 801, 800, like blackmagic recommends here: https://forum.blackmagicdesign.com/viewtopic.php?f=12&t=75482#p418341 ?

**Comment by Julusian** [comment]
@TomKaltz I think the issues are because we currently aren't handling the cadence at all. Some stuff got removed during 2.2 to handle the cadences which https://github.com/CasparCG/server/pull/1326 proposed to reimplement but that stalled.

**Comment by TomKaltz** [comment]
In the code it looks as if the decklink consumer consumes audio packets according to the proper cadence so that’s good.  System audio seems to be the only consumer that isn’t handling fractional frame rates correctly.

**Comment by Julusian** [comment]
yeah some consumers are consuming correctly, but the producing and mixing flow is not correct.

Some producers are producing correctly, but they will sometimes produce with the wrong number of samples, with the mixer either dropping or repeating the last sample to get the 'correct' amount.

Either the producers need to listen be fixed to produce the correct number of samples for the frame about to be mixed (which it looks like was ronag's intention due to `nb_samples` being added to the `receive` methods). Or the mixer needs to buffer samples to handle this misalignment (like it used to do and #1326 wanted to reintroduce)

I don't know why it seems to be working for you. maybe the repeating&dropping is good enough to not notice it most of the time? 
I am curious if commenting out the line https://github.com/CasparCG/server/blob/01aff82f422489acd6f88be8952d54abc7529526/src/core/mixer/audio/audio_mixer.cpp#L109 makes anything noticable from a decklink

**Comment by TomKaltz** [comment]
I would consider the audio on decklink clean and clear when all elements are 1080i5994 (channel, decklink, file).  I may have heard very faint anomalies at times though.  No way to tell for sure as I was testing with a wide variety of source footage.  My method for testing this was to record a casparcg decklink output looped back into media express and then listening to that recorded clip with quicktime player and VLC.  I do still need to try this in a true broadcast environment with genlock.

> I don't know why it seems to be working for you. maybe the repeating&dropping is good enough to not notice it most of the time? I am curious if commenting out the line
> 
> [server/src/core/mixer/audio/audio_mixer.cpp](https://github.com/CasparCG/server/blob/01aff82f422489acd6f88be8952d54abc7529526/src/core/mixer/audio/audio_mixer.cpp#L109)
> 
> Line 109 in [01aff82](/CasparCG/server/commit/01aff82f422489acd6f88be8952d54abc7529526)
> 
>  mixed[n]    = static_cast<double>(ptr[offset]) * item.transform.volume + mixed[n]; 
> makes anything noticable from a decklink

@Julusian doing the above definitely breaks things and results in horrible audio garble on decklink output.

**Comment by TomKaltz** [comment]
@Julusian Is there a way to get logging when samples are dropped or duplicated?

**Comment by Julusian** [comment]
> Is there a way to get logging when samples are dropped or duplicated?

Not that I can see. I suppose something could be added fairly easily, but until this is fixed it will produce constant noise, and hopefully will become unnecessary once it has been fixed.

I am trying to figure out what @ronag's idea was on how to do this better than before. I would guess it is something related to the `nb_samples` parameter being given to producers when pulling frames, but I don't know how that could work cleanly.

Maybe the producers are then supposed to maintain separate buffers of audio and video, only stitching them together inside receive once we know how many samples to attach?
That doesn't sound too horrible to handle in most cases, except for the various places where we call receive asking for 0 samples (a common implementation of last_frame)

And any producer which layers other producers (route and transitions), they will need to do this same kind of buffering of their inputs to ensure they produce the correct samples. For the route producer this would probably be easiest to do by mixing the audio into a single flattened stream (that would also simplify crossing formats). 

For transitions, they need to not be buffering any samples, otherwise when the transition ends the audio will jump.
Perhaps both challenges here could be solved by a new `peek_frame` method (to replace first_frame`). This would keep it possible for a layer to get the 'first available frame' of a producer, and allow transitions to check if a producer has a frame ready. Which should allow them to avoid buffering any frames.  
Either way this will require a careful review of every path a frame can take out of a producer and through other producers, there are other paths which might push us to the separate methods (eg `last_frame` commonly calls receive asking for 0 samples)  

Another patch of fiddly logic I can see is to avoid drift when dropping frames. eg, if receive is called with 800 801 800, and the 801 call doesnt produce a frame, then we have drifted by a sample. repeat this a every few minutes over a day long stream and it could add up. 

Having a quick look at some other software, a common approach is to have different buffers/streams for the video and audio. I am not keen on doing that, it would be a very large change and would likely introduce a series of bugs about sync issues.

@ronag, if you see this then some input from would be appreciated, otherwise I am very tempted to do something along the lines of #1326. While it may not be the best solution, it worked (as far as Im aware) and is better than nothing.

**Comment by ronag** [comment]
I think I already did this once in some branch. Will need to review that code.

**Comment by TomKaltz** [comment]
I hacked [this together](https://github.com/TomKaltz/Server/commit/5b59bd0f84514d409e2ab398e8190236ee2ec8b0)  based on #1326 and in my quick testing seems to fix the issue on decklink and oal output.  @ronag is there anything inherently wrong with this approach?

**Comment by TomKaltz** [comment]
I've updated PR #1615 and these issues seem to be fixed.  [Here are windows binaries](https://drive.google.com/file/d/1kvZjX_KoIw1vjCl4SsBJM5EwMrHmUjAM/view?usp=sharing) if anyone in 5994 land wants to give b17f35ed00c93aad3b54841a5e56450bff498eca of that PR branch a test drive.

**Comment by BonjwaMaurice** [comment]
I’ve tested the build from PR #1615 (commit b17f35e) and can confirm that the audio output at 1080p5994 is now clean. The previously reported issue with crackling and sample drops/repeats no longer occurs.
